doc/fluid/design/ir/draft.md

```diff
@@ -0,0 +1,81 @@
+## Motivation
+
+There is a ```gap``` between the ```Program``` defined by
+user and the ```Executable``` that can be scheduled
+efficiently on heterogeneous hardware, either locally
+or distributedly.
+
+Usually, the ```gap``` is bridged by
+
+* A serious transformations with defined order.
+
+* The transformations usually invovle
+```insert, delete, clustering, split, dependency analysis```.
+
+* Has a simple way to verify and debug each transformation.
+
+* Flexible to add, remove or customize transformations to fit
+the requirements of various algorithms (models) and hardware secenarios.
+
+Some other events also push us to a better unified pattern.
+
+* The deep learning framework is built around the concepts of graphs.
+To leverage tools such as compilation (e.g. TVM and nGraph) or
+cross-framework conversion (e.g. ONNX), we also need a intermediate
+representation that can be connected to the rest of the ecosystem.
+
+
+We need a unified pattern to naturally support the requirements
+described above. The pattern should fit both training, inference
+and other offline serielized model transformations.
+Learned from LLVM and other deep learning framework, we draft the
+design below.
+
+
+## Design
+
+### Major Concepts
+
+#### Node
+
+```Node``` represents an operation that performs some computation or
+a variable that is input or output of operation.
+
+```Node```s are connected to other ```Node```s via inputs and outputs.
+
+#### Graph
+
+```Graph``` contains a list of ```Node```s.
```

Comment:
Thanks for the awesome PR! I have some question about Graph:
1. Do the nodes in a graph are connected? How to determinate the `computation` user defined in which node?
1. How to determinate the execution order of the nodes?

Reply:
1. Currently, there's no enforcement that the nodes in the graph should all be connected. For distributed training, one idea is to have the DistributedTrainPass generating different subgraph in trainer and ps. For example. There might be other ways though.
```
Pass::Apply(graph) {
  if (trainer_): return GetTrainGraph(graph);
  else: return GetPSGraph()
}
```

2. Operation nodes with all dependencies satisfied with the executed first. By default, the graph must have some nodes with no inputs or inputs ready (fed in). After these ready nodes are done and generate new outputs, more dependencies will be satisfied and more nodes will be triggered. Eventually, all nodes in the graph are executed.
I'm also considering running only part of the graph based on "fetch", however, this might be tricky given the program design.